src/net/http/UploadFile.cpp

```diff
@@ -0,0 +1,253 @@
+  return 0;
+}
+
+//-----------------------------------------------------------------------------
+static struct curl_httppost* formpost = nullptr;
+static struct curl_httppost* lastptr = nullptr;
+
+//-----------------------------------------------------------------------------
+static void
+UploadFile(CurlGlobal &curl, const TCHAR *url,
+               const TCHAR* username, const TCHAR*password,
+               void* _buffer, size_t max_length,
+               OperationEnvironment &env) {
+  assert(url != nullptr);
+
+  UploadFileHandler handler(_buffer, max_length, env);
+  CurlRequest request(curl, url, handler);
+  request.SetOption(CURLOPT_HTTPPOST, formpost);
+ 
+  request.SetProgress((void*)OnProgress, &handler);
+  AtScopeExit(&request) { request.StopIndirect(); };
+
+  request.SetFailOnError();
+
+  if (username != nullptr)
+    request.SetOption(CURLOPT_USERNAME, username);
+  if (password != nullptr)
+    request.SetOption(CURLOPT_PASSWORD, password);
+
+  env.SetCancelHandler([&]{
+    request.StopIndirect();
+    handler.Cancel();
+  });
+
+  AtScopeExit(&env) { env.SetCancelHandler({}); };
+
+  request.StartIndirect();
+  handler.Wait();
+}
+
+//-----------------------------------------------------------------------------
+void
+Net::UploadFile(CurlGlobal &curl, const TCHAR *url,
+                    const TCHAR *username, const TCHAR *password,
+                    Path path, char* buffer, size_t buffer_size,
+                    OperationEnvironment &env)
+{
+  assert(url != nullptr);
+  assert(path != nullptr);
+
+  ::UploadFile(curl, url, username, password, buffer, buffer_size, env);
+
+}
+
+//-----------------------------------------------------------------------------
+Net::UploadFileJob::UploadFileJob(CurlGlobal& _curl, const TCHAR * _url, Path _path,
+  char* _buffer, size_t _buffer_size)
+  :curl(_curl), url(_url), path(_path), buffer(_buffer), buffer_size(_buffer_size)
+{
+  NarrowString<MAX_PATH> filename;
+  filename.SetASCII(path.c_str());
+  LogFormat("UploadFile: %s", filename.c_str());
+  curl_formfree(formpost);
+  curl_formadd(&formpost,
+    &lastptr,
+    CURLFORM_COPYNAME, "file",
+    CURLFORM_FILE, filename.c_str(),
+    CURLFORM_CONTENTTYPE, "application/octet-stream",
+    CURLFORM_END);
+}
+
+//-----------------------------------------------------------------------------
+void
+Net::UploadFileJob::Run(OperationEnvironment &env)
+{
+  UploadFile(curl, url, username, password, path, buffer, buffer_size, env);
+}
+
+//-----------------------------------------------------------------------------
+void
+Net::UploadFileJob::AddValue(const char* name, const char* value) {
+  curl_formadd(&formpost,
+    &lastptr,
+    CURLFORM_COPYNAME, name,
+    CURLFORM_COPYCONTENTS, value,
+    CURLFORM_END);
+}
+
+//-----------------------------------------------------------------------------
+#ifdef UNICODE
+void
+Net::UploadFileJob::AddValue(const char* name, const TCHAR* _value) {
+  NarrowString<MAX_PATH> value;
```

Comment:
This means only char buffers allowed? And convertions are doing outside?
What is your prefered solution for this convertion? I found some different, but  want to use the 'favorit'! Maybe like so:
```
  NarrowString<MAX_PATH> char_string;
  char_string.SetASCII(path.c_str());
```
Or do you have a better proposal?

Reply:
This code sample will break everything that's not plain ASCII. That can sometimes be OK, but most of the time not. You didn't explain the context.

Reply:
I want to convert a TCHAR* string, a StaticString or a Path to a plain ASCII char string. It seems the XCSoar has a lot of possibilities to do that, but I want to know, what is the most prefered solution for such task, otherwise we lose a lot of time to try everything...

Reply:
There is no general solution which fits everything. Context matters.
`TCHAR*` represents strings displayed to the user and path names. If you want to convert to `char*`, you need to consider the context - what are you targetting? What encoding does the target expect? That needs to be a per-case decision.